File: 2019-09-hash-vs-sort/hashvssort.js
////////////////////////////////////////////////////////////////////////////////
// Core

function hash_unique(items) {
    let unique = new Set(items);
    return Array.from(unique);
}

function sort_unique(items) {
    let sorted = items.slice().sort();
    let result = [];
    var prev;
    for (let item of sorted) {
        if (item !== prev) {
            result.push(item);
            prev = item;
        }
    }
    return result;
}

////////////////////////////////////////////////////////////////////////////////
// Benchmarking

// Difference of two process.hrtime results, as a number of seconds (float)
function time_difference_s(t0, t1) {
    let diff_s = t1[0] - t0[0];
    let diff_ns = t1[1] - t0[1];
    return diff_s + diff_ns / 1e9;
}

function generate_test_items(n_total, unique_ratio) {
    let n_unique = Math.floor(n_total * unique_ratio);
    let unique_items = new Set();
    while (unique_items.size < n_unique) {
        unique_items.add(Math.floor(Math.random() * Math.pow(2, 32) - Math.pow(2, 31)));
    }
    let items = Array.from(unique_items);
    for (var i = 0; i < n_total - n_unique; ++i) {
        items.push(items[Math.floor(Math.random() * n_unique)]);
    }
    return items;
}

console.log("n_total,n_unique,method,time");
for (var rep = 0; rep < 30; ++rep) {
    for (let log_total of [16, 18, 20, 22]) {
        for (let unique_ratio of [1/16, 1/2, 1]) {
            let items = generate_test_items(1 << log_total, unique_ratio);
            for (let uniquer of [hash_unique, sort_unique]) {
                let start = process.hrtime();
                let unique = uniquer(items);
                let elapsed = time_difference_s(start, process.hrtime());
                console.log(`${items.length},${unique.length},${uniquer.name},${elapsed}`);
            }
        }
    }
}

File: 2019-09-hash-vs-sort/HashVsSort.java
import java.util.Arrays;
import java.util.Collection;
import java.util.function.Function;
import java.util.HashSet;
import java.util.Random;


public class HashVsSort {
    ////////////////////////////////////////////////////////////////////////////////
    // Core

    private static int[] toIntArray(Collection<Integer> collection) {
        int[] results = new int[collection.size()];
        int index = 0;
        for (Integer value : collection) {
            results[index++] = value;
        }
        return results;
    }

    private static int[] hashUnique(int[] items) {
        HashSet<Integer> unique = new HashSet<>();
        for (int item : items) {
            unique.add(item);
        }
        return toIntArray(unique);
    }

    private static int[] sortUnique(int[] items) {
        if (items.length == 0) {
            return items;
        }
        int[] sorted = Arrays.copyOf(items, items.length);
        Arrays.sort(sorted);
        int writeIndex = 1;
        for (int i = 1; i < sorted.length; ++i) {
            if (sorted[i] != sorted[i-1]) {
                sorted[writeIndex++] = sorted[i];
            }
        }
        return Arrays.copyOf(sorted, writeIndex);
    }

    ////////////////////////////////////////////////////////////////////////////////
    // Benchmark

    private static int[] generateTestItems(int nTotal, float uniqueRatio, Random random) {
        int nUnique = (int)(nTotal * uniqueRatio);
        HashSet<Integer> unique = new HashSet<>();
        while (unique.size() < nUnique) {
            unique.add(random.nextInt());
        }
        int[] items = Arrays.copyOf(toIntArray(unique), nTotal);
        for (int i = nUnique; i < nTotal; ++i) {
            items[i] = items[random.nextInt(nUnique)];
        }
        return items;
    }

    private static void runTest(Function<int[], int[]> uniquer, String methodName, int[] items) {
        long start = System.nanoTime();
        int[] unique = uniquer.apply(items);
        long elapsed = System.nanoTime() - start;
        System.out.println(String.format("%d,%d,%s,%f", items.length, unique.length, methodName, elapsed / 1e9));
    }

    public static void main(String[] args) {
        Random random = new Random();
        System.out.println("n_total,n_unique,method,time");
        for (int rep = 0; rep < 30; ++rep) {
            for (int logTotal : Arrays.asList(16, 18, 20, 22)) {
                for (float uniqueRatio : Arrays.asList(1/16.0f, 1/2.0f, 1.0f)) {
                    int[] items = generateTestItems(1 << logTotal, uniqueRatio, random);
                    // Swap order each rep, for fairness
                    if (rep % 2 == 0) {
                        runTest(HashVsSort::hashUnique, "hash_unique", items);
                        runTest(HashVsSort::sortUnique, "sort_unique", items);
                    } else {
                        runTest(HashVsSort::sortUnique, "sort_unique", items);
                        runTest(HashVsSort::hashUnique, "hash_unique", items);
                    }
                }
            }
        }
    }
}
